dissect/target/containers/fortifw.py

```diff
@@ -0,0 +1,206 @@
+import gzip
+import io
+import logging
+import os
+import zlib
+from itertools import cycle, islice
+from pathlib import Path
+from typing import BinaryIO, Optional, Sequence, Union
+
+from dissect.util.stream import RangeStream, RelativeStream
+
+from dissect.target.container import Container
+from dissect.target.tools.utils import catch_sigpipe
+
+logger = logging.getLogger(__name__)
+
+
+def find_xor_key(fobj: io.BytesIO) -> bytes:
+    """Find the XOR key for the firmware file by using known plaintext of zeros.
+
+    File object ``fobj`` should be at the correct offset where it should decode to all zeroes (0x00).
+
+    Arguments:
+        fobj: file object to read from
+
+    Returns:
+        bytes: XOR key, zero bytes if no key is found
+    """
+    key = bytearray()
+
+    pos = fobj.tell()
+    buf = fobj.read(32)
+    fobj.seek(pos)
+
+    if pos % 512 == 0:
+        xor_char = 0xFF
+    else:
+        fobj.seek(pos - 1)
+        xor_char = ord(fobj.read(1))
+
+    for i, k_char in enumerate(buf):
+        idx = (i + pos) & 0x1F
+        key.append((xor_char ^ k_char ^ idx) & 0xFF)
+        xor_char = k_char
+
+    # align xor key
+    koffset = 32 - (pos & 0x1F)
+    key = islice(cycle(key), koffset, koffset + 32)
+    return bytes(key)
+
+
+class FortiFirmwareFile:
+    """Fortinet firmware file, handles transparant decompression and deobfuscation of the firmware file."""
+
+    def __init__(self, fobj: io.BytesIO) -> None:
+        self.fh = fobj
+        self.size = None
+
+        # Check if the file is gzipped
+        self.is_gzipped = False
+        self.fh.seek(0)
+        header = self.fh.read(4)
+        if header.startswith(b"\x1f\x8b"):
+            self.is_gzipped = True
+
+            # Find the extra metadata behind the gzip compressed data
+            # as a bonus we can also calculate the size of the firmware here
+            dec = zlib.decompressobj(wbits=16 + zlib.MAX_WBITS)
+            self.fh.seek(0)
+            self.size = 0
+            while True:
+                data = self.fh.read(io.DEFAULT_BUFFER_SIZE)
+                if not data:
+                    break
+                d = dec.decompress(dec.unconsumed_tail + data)
+                self.size += len(d)
+
+            # Ignore the trailer data of the gzip file if we have any
+            if dec.unused_data:
+                self.fh.seek(-len(dec.unused_data), os.SEEK_END)
+                self.trailer_offset = self.fh.tell()
```

Comment:
Can you initialize these class attributes to `None` if the file is not compressed?

Reply:
done